src/java/org/apache/cassandra/index/sai/disk/v1/keystore/KeyLookup.java

```diff
@@ -197,43 +197,34 @@ public long clusteredSeekToKey(ByteComparable key, long startingPointId, long en
             updateCurrentBlockIndex(startingPointId);
             resetToCurrentBlock();
 
-            if (compareKeys(currentKey, skipKey) == 0)
-                return startingPointId;
+            // We can return straightaway if the currentPointId is within the requested partition range and the keys match
+            if (currentPointId >= startingPointId && currentPointId < endingPointId && compareKeys(currentKey, skipKey) == 0)
+                return currentPointId;
 
-            if (notInCurrentBlock(startingPointId, skipKey))
-            {
-                long split = (endingPointId - startingPointId) >>> blockShift;
-                long splitPointId = startingPointId;
-                while (split > 0)
-                {
-                    updateCurrentBlockIndex(Math.min((splitPointId >>> blockShift) + split, blockOffsets.length() - 1));
-                    resetToCurrentBlock();
+            long splitStartId = startingPointId;
+            long splitEndId = endingPointId;
 
-                    if (currentPointId >= endingPointId)
-                    {
-                        updateCurrentBlockIndex((endingPointId - 1));
-                        resetToCurrentBlock();
-                    }
-
-                    int cmp = compareKeys(currentKey, skipKey);
-
-                    if (cmp == 0)
-                        return currentPointId;
+            while ((splitEndId - splitStartId) >>> blockShift > 0)
+            {
+                long midPointId = splitStartId + (splitEndId - splitStartId) / 2;
 
-                    if (cmp < 0)
-                        splitPointId = currentPointId;
+                int position = keyInBlockComparison(midPointId, skipKey);
```

Comment:
So essentially this `keyInBlockComparison()` method is `moveToBlockAndCompareTo()`? Going back and forth on whether it would be less surprising to just pull up the `updateCurrentBlockIndex()` and `resetToCurrentBlock()` calls to this level to avoid hiding the "move" part in `keyInBlockComparison()`...

Reply:
I like `moveToBlockAndCompareTo`, so have changed it to that because it is closer to the behaviour. I wanted the update and reset block in this method because it shouldn't really be used without it. The name makes the binary search clearer as well.